projects/components/src/download-file/download-file.component.ts

```diff
@@ -1,67 +1,41 @@
 import { IconType } from '@hypertrace/assets-library';
-import { catchError, finalize, take } from 'rxjs/operators';
+import { Observable } from 'rxjs';
+import { map } from 'rxjs/operators';
 import { ButtonSize, ButtonStyle } from '../button/button';
 import { IconSize } from '../icon/icon-size';
-import { NotificationService } from '../notification/notification.service';
 import { DownloadFileMetadata } from './download-file-metadata';
+import { FileDownloadEventType, FileDownloadService } from './service/file-download.service';
 
 @Component({
   selector: 'ht-download-file',
   changeDetection: ChangeDetectionStrategy.OnPush,
   styleUrls: ['./download-file.component.scss'],
+  providers: [FileDownloadService],
   template: `
     <div *ngIf="this.metadata" class="download-file" (click)="this.triggerDownload()">
       <ht-button
-        *ngIf="!this.dataLoading"
+        *ngIf="!(this.dataLoading$ | async)"
         class="download-button"
         icon="${IconType.Download}"
         display="${ButtonStyle.Text}"
         size="${ButtonSize.Large}"
       ></ht-button>
-      <ht-icon *ngIf="this.dataLoading" icon="${IconType.Loading}" size="${IconSize.Large}"></ht-icon>
+      <ht-icon *ngIf="this.dataLoading$ | async" icon="${IconType.Loading}" size="${IconSize.Large}"></ht-icon>
     </div>
   `
 })
 export class DownloadFileComponent {
   @Input()
   public metadata?: DownloadFileMetadata;
 
-  public dataLoading: boolean = false;
-  private readonly dlJsonAnchorElement: HTMLAnchorElement;
+  public constructor(private readonly fileDownloadService: FileDownloadService) {}
 
-  public constructor(
-    @Inject(DOCUMENT) private readonly document: Document,
-    private readonly renderer: Renderer2,
-    private readonly changeDetector: ChangeDetectorRef,
-    private readonly notificationService: NotificationService
-  ) {
-    this.dlJsonAnchorElement = this.document.createElement('a');
+  public get dataLoading$(): Observable<boolean> {
+    return this.fileDownloadService.fileDownloadEvent$.pipe(map(event => event === FileDownloadEventType.Progress));
   }
 
   public triggerDownload(): void {
-    this.dataLoading = true;
-    this.metadata!.dataSource.pipe(
-      take(1),
-      catchError(() => this.notificationService.createFailureToast('Download failed')),
-      finalize(() => {
-        this.dataLoading = false;
-        this.changeDetector.detectChanges();
-      })
-    ).subscribe((data: string) => {
-      this.downloadData(data);
-    });
-  }
-
-  private downloadData(data: string): void {
-    this.renderer.setAttribute(
-      this.dlJsonAnchorElement,
-      'href',
-      `data:text/plain;charset=utf-8,${encodeURIComponent(data)}`
-    );
-    this.renderer.setAttribute(this.dlJsonAnchorElement, 'download', this.metadata!.fileName);
-    this.renderer.setAttribute(this.dlJsonAnchorElement, 'display', 'none');
-    this.dlJsonAnchorElement.click();
+    this.fileDownloadService.downloadAsText(this.metadata!);
```

Comment:
Shouldn't be using `!`. you could pass in the defined `metadata` from the template. That way it will always be defined

Reply:
I agree but my question here is doing the assignation in the constructor (just to learn)

Reply:
> my question here is doing the assignation in the constructor (just to learn)

`metadata` is an Input field so we can't assign anything in the constructor. It will be `undefined`

Reply:
So you can use OnChanges or in the worst case on OnInit

Reply:
Yeah, we can. As we're already doing ngIf in the template, Anand's method will also work. We can choose any of these